src/radical/pilot/pmgr/launching/default.py

```diff
@@ -1336,5 +1199,134 @@ def _prepare_pilot(self, resource, rcfg, pilot, expand):
         return ret
 
 
+    # --------------------------------------------------------------------------
+    #
+    def _stage_in(self, pilot, sds):
+        '''
+        Run some input staging directives.
+        '''
+
+        resource_sandbox = self._session._get_resource_sandbox(pilot)
+        session_sandbox  = self._session._get_session_sandbox (pilot)
+        pilot_sandbox    = self._session._get_pilot_sandbox   (pilot)
+        client_sandbox   = self._session._get_client_sandbox()
+
+        # contexts for staging url expansion
+        rem_ctx = {'pwd'     : pilot_sandbox,
+                   'client'  : client_sandbox,
+                   'pilot'   : pilot_sandbox,
+                   'resource': resource_sandbox}
+
+        loc_ctx = {'pwd'     : client_sandbox,
+                   'client'  : client_sandbox,
+                   'pilot'   : pilot_sandbox,
+                   'resource': resource_sandbox}
+
+        sds = ru.as_list(sds)
+
+        for sd in sds:
+            sd['prof_id'] = pilot['uid']
+
+        for sd in sds:
+            sd['source'] = str(complete_url(sd['source'], loc_ctx, self._log))
+            sd['target'] = str(complete_url(sd['target'], rem_ctx, self._log))
+
+        self._stage(sds)
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _stage_out(self, pilot, sds):
+        '''
+        Run some output staging directives.
+        '''
+
+        resource_sandbox = self._session._get_resource_sandbox(pilot)
+        session_sandbox  = self._session._get_session_sandbox (pilot)
+        pilot_sandbox    = self._session._get_pilot_sandbox   (pilot)
+        client_sandbox   = self._session._get_client_sandbox()
+
+        # contexts for staging url expansion
+        loc_ctx = {'pwd'     : pilot_sandbox,
+                   'client'  : client_sandbox,
+                   'pilot'   : pilot_sandbox,
+                   'resource': resource_sandbox}
+
+        rem_ctx = {'pwd'     : client_sandbox,
+                   'client'  : client_sandbox,
+                   'pilot'   : pilot_sandbox,
+                   'resource': resource_sandbox}
+
+        sds = ru.as_list(sds)
+
+        for sd in sds:
+            sd['prof_id'] = pilot['uid']
+
+        for sd in sds:
+            sd['source'] = str(complete_url(sd['source'], self._rem_ctx, self._log))
```

Comment:
```
sd['source'] = str(complete_url(sd['source'], loc_ctx, self._log))
sd['target'] = str(complete_url(sd['target'], rem_ctx, self._log))
```

may be change variables `loc_ctx` and `rem_ctx` into `src_ctx` and `tgt_ctx`? (wherever it is used, for less confusion)

Reply:
This will remain confusing: on input staging, the src is local and tgt is remote; on output staging the src is remote and tgt is local...

Reply:
agree with explanation, BUT in the code, for example with output staging, src (as remote) is set as `loc_ctx`, that what I mean with confusion, kinda better just set direction when these variables are defined (not when they are used later)

Reply:
BTW here are local variables (not class/instance related), thus no `self._` and should be switched, and for local variable could be applied what I propose

Reply:
ack on the `self` (fixed) - but the `remote` reference` to the `source` part is correct AFAICS - I don't see `loc_ctx` being used for the source?